Reqnroll/Infrastructure/TestExecutionEngine.cs

```diff
@@ -162,6 +162,7 @@ public virtual async Task OnFeatureEndAsync()
             _testThreadExecutionEventPublisher.PublishEvent(new FeatureFinishedEvent(FeatureContext));
 
             _contextManager.CleanupFeatureContext();
+            _testRunnerManager.ReleaseTestThreadContext(TestThreadContext);
```

Comment:
For the naming:
If we could add this as a breaking change. I would also have added a `Release` or `Dispose` (with `IDisposable`) method. I also added a new method to `TestRunnerManager` but not static. Is this okay too?

For the location of the call to the `Release` method:
Yes the generator needs to call this method explicitly. At least if we support parallel execution of scenarios. 😁 

For the FeatureStart/FeatureEnd methods I'm not that sure if we still need an explicit `TestRunner` instance for them (for example cached in a static like today for MsBuild).
Reasoning: The user can request an `ITestRunner` interface in his `BeforeFeature` hook. He can store this instance (in a static variable) and reuse it later (for example in the `AfterFeature` hook).
Internally we at least need to know when we need to dispose of the `FeatureContext`, so we will probably need it too.
The counter-argument is that we would call the `Release` method differently for `ITestRunner` at Feature-level (implicitly in Featureend) and later at Scenario-Level (explicitly in generated code).
That's why I was fine to add it here (for now). If you want I can add it to the generator and remove it here. But then we will have it in 3 different places (3 generators). 😉 
⚖️

Reply:
Naming:
> If we could add this as a breaking change. I would also have added a `Release` or `Dispose` (with `IDisposable`) method. I also added a new method to TestRunnerManager but not static. Is this okay too?

Hm... OK. Probably if we only *add* new stuff for the `ITestRunner` that is still OK. Hopefully there are no plugins that override `ITestRunner`. If they just use it, it will be still compatible. Whether we make a `Release` or implement  `IDisposable`, I don't really mind. Probably the `IDisposable` is cleaner and will be easier to implement in code gen (with using), once the test runner will be local to the test method.

I understand the reasoning, but I would still vote to move it out and add it (in all three generators, yeah...). I cannot really explain, but somehow I have a fear that there are plugins who use the test runner and call the OnFeatureEnd without expecting that it will automatically end the entire test thread and doing anything else afterwards will cause troubles. Also that might guide the other improvements (e.g. roslyn gen) more to the right direction.

Reply:
> I understand the reasoning, but I would still vote to move it out and add it (in all three generators, yeah...).

No problem I can add it to the generators. 🙂 

> Probably if we only add new stuff for the ITestRunner that is still OK. Hopefully there are no plugins that override ITestRunner.

I found one codebase that uses Reqnroll and implement `ITestRunner` them self ([allure-csharp](https://github.com/allure-framework/allure-csharp/blob/bd04453cb1276beb29eeebbc32d8d5c6acad05a5/Allure.Reqnroll/SelectiveRun/TestPlanAwareTestRunner.cs#L12)). 😁 

So I'm unsure if we want to add the method or not for now.
If not the Generator would need to do generate something like `testRunner.TestThreadContext.TestThreadContainer.Resolve<ITestRunnerManager>().ReleaseTestThreadContext(testRunner.TestThreadContext);`.

I'm open for both directions. 😉

Reply:
Thx for the research. In this case we should not add anything to the `ITestRunner` but go for the other option. Generating such a complex c# expression is hard (generic method calls), so feel free to create a static method on the `TestRunnerManager` to do all this (takes testRunner as param).

Reply:
Thanks for the Tip. I changed it that way. 🙂